include/openssl/hpke.h

```diff
@@ -0,0 +1,478 @@
+# define OSSL_HPKE_KEMSTR_X448        SN_X448              /**< KEM id 0x21 */
+# define OSSL_HPKE_KDFSTR_256         "hkdf-sha256"        /**< KDF id 1 */
+# define OSSL_HPKE_KDFSTR_384         "hkdf-sha384"        /**< KDF id 2 */
+# define OSSL_HPKE_KDFSTR_512         "hkdf-sha512"        /**< KDF id 3 */
+# define OSSL_HPKE_AEADSTR_AES128GCM  LN_aes_128_gcm       /**< AEAD id 1 */
+# define OSSL_HPKE_AEADSTR_AES256GCM  LN_aes_256_gcm       /**< AEAD id 2 */
+# define OSSL_HPKE_AEADSTR_CP         LN_chacha20_poly1305 /**< AEAD id 3 */
+# define OSSL_HPKE_AEADSTR_EXP        "exporter"           /**< AEAD id 0xff */
+
+/**
+ * @brief ciphersuite combination
+ */
+typedef struct {
+    uint16_t    kem_id; /**< Key Encryption Method id */
+    uint16_t    kdf_id; /**< Key Derivation Function id */
+    uint16_t    aead_id; /**< AEAD alg id */
+} OSSL_HPKE_SUITE;
+
+/**
+ * Suite constants, use this like:
+ *          OSSL_HPKE_SUITE myvar = OSSL_HPKE_SUITE_DEFAULT;
+ */
+# define OSSL_HPKE_SUITE_DEFAULT \
+    {\
+        OSSL_HPKE_KEM_ID_X25519, \
+        OSSL_HPKE_KDF_ID_HKDF_SHA256, \
+        OSSL_HPKE_AEAD_ID_AES_GCM_128 \
+    }
+
+# ifndef OSSL_HPKE_MAXSIZE
+#  define OSSL_HPKE_MAXSIZE 512
+# endif
+
+/**
+ * @brief opaque type for HPKE contexts
+ */
+typedef struct ossl_hpke_ctx_st OSSL_HPKE_CTX;
+
+/**
+ * @brief context creator
+ * @param mode is the desired HPKE mode
+ * @param suite specifies the KEM, KDF and AEAD to use
+ * @param libctx is the library context to use
+ * @param propq is a properties string for the library
+ * @return pointer to new context or NULL if error
+ */
+OSSL_HPKE_CTX *OSSL_HPKE_CTX_new(int mode, OSSL_HPKE_SUITE suite,
+                                 OSSL_LIB_CTX *libctx, const char *propq);
+
+/**
+ * @brief free up storage for a HPKE context
+ * @param ctx is the pointer to be free'd (can be NULL)
+ */
+void OSSL_HPKE_CTX_free(OSSL_HPKE_CTX *ctx);
+
+/**
+ * @brief set a PSK for an HPKE context
+ * @param ctx is the pointer for the HPKE context
+ * @param pskid is a string identifying the PSK
+ * @param psk is the PSK buffer
+ * @param psklen is the size of the PSK
+ * @return 1 for success, 0 for error
+ */
+int OSSL_HPKE_CTX_set1_psk(OSSL_HPKE_CTX *ctx,
+                           const char *pskid,
+                           const unsigned char *psk, size_t psklen);
+
+/**
+ * @brief set a sender KEM private key for HPKE
+ * @param ctx is the pointer for the HPKE context
+ * @param privp is an EVP_PKEY form of the private key
+ * @return 1 for success, 0 for error
+ *
+ * If no key is set via this API an ephemeral one will be
+ * generated in the first seal operation and used until the
+ * context is free'd. (Or until a subsequent call to this
+ * API replaces the key.) This suits senders who are typically
+ * clients.
+ */
+int OSSL_HPKE_CTX_set1_senderpriv(OSSL_HPKE_CTX *ctx, EVP_PKEY *privp);
+
+#ifdef OSSL_KEM_PARAM_OPERATION_DHKEM
```

Comment:
This #ifdef  should be removed...
It should be documented with a warning.. I need to add the warning to the DHKEM documentation also.

Reply:
I am assuming the main use of this API is for testing..

Reply:
wrt testing yeah, and it could be considered a bit of a dangerous implement; OTOH, the deterministic ikm keygen has the same potential for deliberate bad actors (generate known keys) so don't think it's incrementally more dangerous and it could be useful for some test vectors that'd otherwise be hard to do - overall though, yes this could be dropped without loss of function (I'm fine with doing so if that's best)

Reply:
OK, so I've figured this out and there's a decision to make. As defined, (and it's v. unlikely to change now there are some crypto proofs for ECH) ECH requires that the AAD for a seal operation include the sender's ephemeral public key (``enc``). That in turn means we need to generate the key pair before doing the seal operation. I've got a build with working ECH that generates random initial key material (IKM) and then uses ``OSSL_HPKE_sender_seal()`` (and hence the DHKEM APIs) with that IKM so we end up with the same ``enc`` value that the recipient will use as part of the AAD for the ``OSSL_HPKE_recipient_open()`` operation. That works but is non-elegant:-) (Non-elegant, because the ECH code then has to decide how much randomness to generate which is really knowledge that ought be closer to the crypto.)

An alternative would be for DHKEM to allow for re-use of an EVP_PKEY* via an API that maps to ``OSSL_HPKE_set1_senderpriv()`` (IOW, provide a way to hand an EVP_PKEY* to the DHKEM APIs). I've not found a way to do that yet but maybe it exists. It certainly could be added I guess.

So - the choice for now seems to be: a) live with ECH code determining how many random bits of IKM to use (and then get rid of the ``OSSL_HPKE_set1_senderpriv()`` API) or b) provide a way to make ``OSSL_HPKE_set1_senderpriv()`` work when using the DHKEM APIs.

I'm neutral as to which'd be better but will go with (a) since it works if I don't get contrary input.

Reply:
OK, so I've figured this out and there's a decision to make. As defined, (and it's v. unlikely to change now there are some crypto proofs for ECH) ECH requires that the AAD for a seal operation include the sender's ephemeral public key (``enc``). That in turn means we need to generate the key pair before doing the seal operation. I've got a build with working ECH that generates random initial key material (IKM) and then uses ``OSSL_HPKE_sender_seal()`` (and hence the DHKEM APIs) with that IKM so we end up with the same ``enc`` value that the recipient will use as part of the AAD for the ``OSSL_HPKE_recipient_open()`` operation. That works but is non-elegant:-) (Non-elegant, because the ECH code then has to decide how much randomness to generate which is really knowledge that ought be closer to the crypto.)

An alternative would be for DHKEM to allow for re-use of an EVP_PKEY* via an API that maps to ``OSSL_HPKE_set1_senderpriv()`` (IOW, provide a way to hand an EVP_PKEY* to the DHKEM APIs). I've not found a way to do that yet but maybe it exists. It certainly could be added I guess.

So - the choice for now seems to be: a) live with ECH code determining how many random bits of IKM to use (and then get rid of the ``OSSL_HPKE_set1_senderpriv()`` API) or b) provide a way to make ``OSSL_HPKE_set1_senderpriv()`` work when using the DHKEM APIs.

I'm neutral as to which'd be better but will go with (a) since it works if I don't get contrary input.

Reply:
In terms of interface,  the seal just needs an aad and symmetric key.. The seal interface doesnt care what the aad is , as long as its the same value used for both seal and open..

Reply:
overtaken by events